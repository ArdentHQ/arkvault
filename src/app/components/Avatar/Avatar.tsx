import { Helpers } from "@ardenthq/sdk-profiles";
import cn from "classnames";
import React from "react";
import tw, { styled } from "twin.macro";

import { Size } from "@/types";

interface Properties {
	address?: string;
	className?: string;
	innerClassName?: string;
	shadowClassName?: string;
	highlight?: boolean;
	noShadow?: boolean;
	size?: Size;
	children?: React.ReactNode;
}

const AvatarWrapper = styled.div<Properties>`
	${tw`transition-all duration-100 relative inline-flex items-center justify-center align-middle rounded-full`}

	${({ size }) => {
		const sizes = {
			avatarMobile: () => tw`w-[25px] h-[25px] text-sm`,
			default: () => tw`w-10 h-10`,
			lg: () => tw`w-11 h-11 text-sm`,
			sm: () => tw`w-8 h-8 text-sm`,
			xl: () => tw`w-16 h-16 text-xl`,
			xs: () => tw`w-5 h-5 text-sm`,
		};

		// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
		return (sizes[size as keyof typeof sizes] || sizes.default)();
	}}

	${({ noShadow, shadowClassName }) => {
		if (noShadow) {
			return;
		}

		if (shadowClassName) {
			return tw`ring-6`;
		}

		return tw`ring-6 ring-theme-background`;
	}}
`;

export const Avatar = ({
	address = "",
	className,
	innerClassName = "rounded-full",
	highlight,
	noShadow,
	shadowClassName,
	size,
	children,
}: Properties) => {
	const svg = React.useMemo(() => (address ? Helpers.Avatar.make(address) : undefined), [address]);

	return (
		<AvatarWrapper
			data-testid="Avatar"
			size={size}
			noShadow={!!noShadow}
			className={cn(className, shadowClassName, "shrink-0")}
			shadowClassName={shadowClassName}
		>
			<div
				className={cn(
					"inline-flex h-full w-full items-center justify-center overflow-hidden align-middle",
					{ "ring-2 ring-theme-primary-600": highlight },
					innerClassName,
				)}
			>
				{svg && <img alt={address} title={address} src={`data:image/svg+xml;utf8,${svg}`} />}
				{children}
			</div>
		</AvatarWrapper>
	);
};
