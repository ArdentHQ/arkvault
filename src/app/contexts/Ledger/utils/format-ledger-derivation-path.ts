import { LedgerDerivationScheme } from "@/app/contexts/Ledger/Ledger.contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
