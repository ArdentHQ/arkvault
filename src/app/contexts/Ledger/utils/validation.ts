import semver from "semver";
import { Coins } from "@payvo/sdk";
import { minVersionList } from "@/app/contexts/Ledger/Ledger.contracts";

export const hasRequiredAppVersion = async (coin: Coins.Coin) => {
	const coinId = coin.network().coin();

	if (minVersionList[coinId]) {
		const currentVersion = await coin.ledger().getVersion();

		if (semver.lt(currentVersion, minVersionList[coinId])) {
			return false;
		}
	}

	return true;
};
