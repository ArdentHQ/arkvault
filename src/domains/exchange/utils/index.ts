export * from "./assertions";
export * from "./errors";
