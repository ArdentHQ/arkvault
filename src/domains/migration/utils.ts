import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
