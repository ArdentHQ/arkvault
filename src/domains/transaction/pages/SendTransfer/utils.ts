export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
