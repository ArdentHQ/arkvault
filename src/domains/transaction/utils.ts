import { Services } from "@ardenthq/sdk";
import { RecipientItem } from "@/domains/transaction/components/RecipientList/RecipientList.contracts";

export const isNoDeviceError = (error: any) => {
	if (!error) {
		return false;
	}

	if (String(error).includes("no device found")) {
		return true;
	}

	return false;
};

export const isRejectionError = (error: any) => String(error).includes("Condition of use not satisfied");

export const handleBroadcastError = ({ errors }: Services.BroadcastResponse) => {
	const allErrors = Object.values(errors);

	if (allErrors.length === 0) {
		return;
	}

	throw new Error(allErrors[0]);
};

export const getTransferType = ({ recipients }: { recipients: RecipientItem[] }): "multiPayment" | "transfer" => {
	const isMultiPayment = recipients.length > 1;
	return isMultiPayment ? "multiPayment" : "transfer";
};

export const withAbortPromise =
	(signal?: AbortSignal, callback?: () => void) =>
	<T>(promise: Promise<T>) =>
		new Promise<T>((resolve, reject) => {
			if (signal) {
				signal.addEventListener("abort", () => {
					callback?.();
					reject("ERR_ABORT");
				});
			}

			return promise.then(resolve).catch(reject);
		});
