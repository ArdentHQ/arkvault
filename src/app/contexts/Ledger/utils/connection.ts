import { Coins } from "@ardenthq/sdk";
import retry, { AbortError, Options } from "p-retry";

import { formatLedgerDerivationPath } from "./format-ledger-derivation-path";
import { hasRequiredAppVersion } from "./validation";

export const accessLedgerDevice = async (coin: Coins.Coin) => {
	try {
		await coin.__construct();
		await coin.ledger().connect();
	} catch (error) {
		// If the device is open, continue normally.
		// Can be triggered when the user retries ledger connection.
		if (error.message !== "The device is already open.") {
			throw error;
		}
	}
};

export const accessLedgerApp = async ({ coin }: { coin: Coins.Coin }) => {
	await accessLedgerDevice(coin);

	if (!(await hasRequiredAppVersion(coin))) {
		throw new Error("VERSION_ERROR");
	}

	// Ensure that the app is accessible.
	await coin.ledger().getPublicKey(
		formatLedgerDerivationPath({
			coinType: coin.config().get<number>("network.constants.slip44"),
		}),
	);
};

export const persistLedgerConnection = async ({
	coin,
	options,
	hasRequestedAbort,
}: {
	coin: Coins.Coin;
	options: Options;
	hasRequestedAbort: () => boolean;
}) => {
	const retryAccess: any = async (attempts: number) => {
		if (hasRequestedAbort() && attempts > 1) {
			throw new AbortError("CONNECTION_ERROR");
		}

		try {
			await accessLedgerApp({ coin });
		} catch (error) {
			// Abort on version error or continue retrying access.
			if (error.message === "VERSION_ERROR") {
				throw new AbortError("VERSION_ERROR");
			}

			throw error;
		}
	};

	await retry(retryAccess, options);
};
