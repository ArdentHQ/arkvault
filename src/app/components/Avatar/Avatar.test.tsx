import React from "react";

import { Avatar } from "./Avatar";
import { Size } from "@/types";
import { render, screen } from "@/utils/testing-library";

describe("Avatar", () => {
	it("should render", () => {
		const { asFragment } = render(<Avatar address="abc" />);

		expect(screen.getByTestId("Avatar")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with highlight", () => {
		const { asFragment } = render(<Avatar address="abc" highlight />);

		expect(screen.getByTestId("Avatar")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with custom shadow color", () => {
		const { asFragment } = render(<Avatar address="abc" shadowClassName="ring-theme-black" />);

		expect(screen.getByTestId("Avatar")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with highlight and custom shadow color", () => {
		const { asFragment } = render(<Avatar address="abc" shadowClassName="ring-theme-black" highlight />);

		expect(screen.getByTestId("Avatar")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render without shadow", () => {
		const { asFragment } = render(<Avatar address="abc" size="lg" noShadow />);

		expect(screen.getByTestId("Avatar")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it.each(["xs", "sm", "lg", "xl"])("should render with size", (size) => {
		const { asFragment } = render(<Avatar address="abc" size={size as Size} />);

		expect(screen.getByTestId("Avatar")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});
});
