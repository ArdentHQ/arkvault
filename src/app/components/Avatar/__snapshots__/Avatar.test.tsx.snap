// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar > should render 1`] = `
<DocumentFragment>
  <div
    class="shrink-0 css-jfogd3"
    data-testid="Avatar"
  >
    <div
      class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle"
    >
      <img
        alt="abc"
        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(76, 175, 80)" width="100" height="100"/><circle r="45" cx="30" cy="60" fill="rgb(255, 193, 7)"/><circle r="50" cx="50" cy="30" fill="rgb(244, 67, 54)"/><circle r="60" cx="80" cy="40" fill="rgb(0, 150, 136)"/></svg>"
        title="abc"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should render with custom shadow color 1`] = `
<DocumentFragment>
  <div
    class="ring-theme-black shrink-0 css-16qzgxm"
    data-testid="Avatar"
  >
    <div
      class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle"
    >
      <img
        alt="abc"
        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(76, 175, 80)" width="100" height="100"/><circle r="45" cx="30" cy="60" fill="rgb(255, 193, 7)"/><circle r="50" cx="50" cy="30" fill="rgb(244, 67, 54)"/><circle r="60" cx="80" cy="40" fill="rgb(0, 150, 136)"/></svg>"
        title="abc"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should render with highlight 1`] = `
<DocumentFragment>
  <div
    class="shrink-0 css-jfogd3"
    data-testid="Avatar"
  >
    <div
      class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle ring-2 ring-theme-primary-600"
    >
      <img
        alt="abc"
        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(76, 175, 80)" width="100" height="100"/><circle r="45" cx="30" cy="60" fill="rgb(255, 193, 7)"/><circle r="50" cx="50" cy="30" fill="rgb(244, 67, 54)"/><circle r="60" cx="80" cy="40" fill="rgb(0, 150, 136)"/></svg>"
        title="abc"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should render with highlight and custom shadow color 1`] = `
<DocumentFragment>
  <div
    class="ring-theme-black shrink-0 css-16qzgxm"
    data-testid="Avatar"
  >
    <div
      class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle ring-2 ring-theme-primary-600"
    >
      <img
        alt="abc"
        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(76, 175, 80)" width="100" height="100"/><circle r="45" cx="30" cy="60" fill="rgb(255, 193, 7)"/><circle r="50" cx="50" cy="30" fill="rgb(244, 67, 54)"/><circle r="60" cx="80" cy="40" fill="rgb(0, 150, 136)"/></svg>"
        title="abc"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should render with size 1`] = `
<DocumentFragment>
  <div
    class="shrink-0 css-1yk316i"
    data-testid="Avatar"
  >
    <div
      class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle"
    >
      <img
        alt="abc"
        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(76, 175, 80)" width="100" height="100"/><circle r="45" cx="30" cy="60" fill="rgb(255, 193, 7)"/><circle r="50" cx="50" cy="30" fill="rgb(244, 67, 54)"/><circle r="60" cx="80" cy="40" fill="rgb(0, 150, 136)"/></svg>"
        title="abc"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should render with size 2`] = `
<DocumentFragment>
  <div
    class="shrink-0 css-h8w6w0"
    data-testid="Avatar"
  >
    <div
      class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle"
    >
      <img
        alt="abc"
        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(76, 175, 80)" width="100" height="100"/><circle r="45" cx="30" cy="60" fill="rgb(255, 193, 7)"/><circle r="50" cx="50" cy="30" fill="rgb(244, 67, 54)"/><circle r="60" cx="80" cy="40" fill="rgb(0, 150, 136)"/></svg>"
        title="abc"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should render with size 3`] = `
<DocumentFragment>
  <div
    class="shrink-0 css-1n9jllj"
    data-testid="Avatar"
  >
    <div
      class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle"
    >
      <img
        alt="abc"
        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(76, 175, 80)" width="100" height="100"/><circle r="45" cx="30" cy="60" fill="rgb(255, 193, 7)"/><circle r="50" cx="50" cy="30" fill="rgb(244, 67, 54)"/><circle r="60" cx="80" cy="40" fill="rgb(0, 150, 136)"/></svg>"
        title="abc"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should render with size 4`] = `
<DocumentFragment>
  <div
    class="shrink-0 css-e8csh4"
    data-testid="Avatar"
  >
    <div
      class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle"
    >
      <img
        alt="abc"
        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(76, 175, 80)" width="100" height="100"/><circle r="45" cx="30" cy="60" fill="rgb(255, 193, 7)"/><circle r="50" cx="50" cy="30" fill="rgb(244, 67, 54)"/><circle r="60" cx="80" cy="40" fill="rgb(0, 150, 136)"/></svg>"
        title="abc"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should render without shadow 1`] = `
<DocumentFragment>
  <div
    class="shrink-0 css-1px82g5"
    data-testid="Avatar"
  >
    <div
      class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle"
    >
      <img
        alt="abc"
        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(76, 175, 80)" width="100" height="100"/><circle r="45" cx="30" cy="60" fill="rgb(255, 193, 7)"/><circle r="50" cx="50" cy="30" fill="rgb(244, 67, 54)"/><circle r="60" cx="80" cy="40" fill="rgb(0, 150, 136)"/></svg>"
        title="abc"
      />
    </div>
  </div>
</DocumentFragment>
`;
